streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -1848,113 +1907,101 @@ object ZChannel {
-                                 (queueReader >>> channel)
-                                   .toPullIn(scope)
-                                   .flatMap(evaluatePull(_).raceAwait(errorSignal.await.interruptible))
-                               }
-                             childFiber <- permits
-                                             .withPermit(latch.succeed(()) *> raceIOs)
-                                             .forkIn(scope)
-                             _       <- latch.await
-                             errored <- errorSignal.isDone
-                           } yield !errored
+                           pull.flatMap(channel =>
+                             for {
+                               latch <- Promise.make[Nothing, Unit]
+                               raceIOs =
+                                 ZIO.scopedWith { scope =>
+                                   (incoming >>> channel)
+                                     .toPullInAlt(scope)
+                                     .flatMap(evaluatePull)
+                                 }
+                               _ <- permits
+                                      .withPermit(latch.succeed(()) *> raceIOs)
+                                      .fork
+                               _ <- latch.await
+                             } yield ()
+                           )
                          case MergeStrategy.BufferSliding =>
-                           for {
-                             canceler <- Promise.make[Nothing, Unit]
-                             latch    <- Promise.make[Nothing, Unit]
-                             size     <- cancelers.size
-                             _        <- ZIO.when(size >= n)(cancelers.take.flatMap(_.succeed(())))
-                             _        <- cancelers.offer(canceler)
-                             raceIOs =
-                               ZIO.scopedWith { scope =>
-                                 (queueReader >>> channel)
-                                   .toPullIn(scope)
-                                   .flatMap(
-                                     evaluatePull(_)
-                                       .raceAwait(errorSignal.await.interruptible)
-                                       .raceAwait(canceler.await.interruptible)
-                                   )
-                               }
-                             childFiber <- permits
-                                             .withPermit(latch.succeed(()) *> raceIOs)
-                                             .forkIn(scope)
-                             _       <- latch.await
-                             errored <- errorSignal.isDone
-                           } yield !errored
+                           pull.flatMap(channel =>
+                             for {
+                               canceler <- Promise.make[Nothing, Unit]
+                               latch    <- Promise.make[Nothing, Unit]
```

Comment:
Let's use the `unsafe` API here as well. You might need to make `pullStrategy` a function that takes `FiberId` as as a param though and call `forever` on the resulting effect